staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go

```diff
@@ -22,12 +25,12 @@ import (
 
 var (
 	// DefaultMutableFeatureGate is a mutable version of DefaultFeatureGate.
-	// Only top-level commands/options setup and the k8s.io/component-base/featuregate/testing package should make use of this.
+	// Only top-level commands/options setup should make use of this.
 	// Tests that need to modify feature gates for the duration of their test should use:
 	//   defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, <value>)()
-	DefaultMutableFeatureGate featuregate.MutableFeatureGate = featuregate.NewFeatureGate()
+	DefaultMutableFeatureGate featuregate.MutableFeatureGate = featuregate.DefaultMutableFeatureGate
```

Comment:
The usual pattern for a feature check is:
- define a feature somewhere
- automatically add it to k8s.io/apiserver/pkg/util/feature.DefaultFeatureGate in the file that defines the feature
- check state of the feature in DefaultFeatureGate

component-base currently can't do that because it has no access to k8s.io/apiserver/pkg/util/feature (circular dependency problem). The first commit on this PR solves that by moving the the FeatureGate instance into component-base/featuregate.

This particular change here ensures that users of k8s.io/apiserver/pkg/util/feature continue to work as before.

I've tried some alternatives for this (see links under "Special notes for your reviewer") but IMHO they are worse.

Reply:
I expected to pass in the instance, not move the global instance to component-base

Reply:
That is how I implemented it in https://github.com/pohly/kubernetes/commit/8d834839f87959a556bdfda8b9e23b5f81b816ab

It makes the API more complicated and I still don't see the advantage of doing it that way. Why should in-tree packages and other staging packages like k8s.io/cloud-provider be allowed to check a global instance, but not component-base? That seems to derive simply from how the code is currently organized, not because it's best for developers or users.

Reply:
because component-base is not about a _particular_ component, and putting the global feature gate there will mean registering feature flags for particular components into what is supposed to be a component-neutral package

Reply:
`k8s.io/apiserver/pkg/util` is also not about any particular component. It just happens to be shared by them.

I'd like to point out one thing: I am not suggesting to move the feature registration into component-base. Someone who consumes component-base/featuregates and nothing else gets an empty component-base/featuregates.DefaultFeatureGate instance.

Features get added to that only when a component also imports the code that defines and expects a certain feature. For component-base/logs that would be the new LoggingAlpha/BetaOptions features and nothing else.

Reply:
If a consumer of component-base/featuregates wants an empty FeatureGate, they can construct one, just as they do now. They are not forced to use the default one.

Reply:
To recap the SIG Architecture discussion that we just had: using a global FeatureGate instance is seen as a problematic code pattern that shouldn't be repeated in component-base if it can be avoided. So it's not about the location of that DefaultFeatureGate instance, it is about what the desired API of component-base/logs and component-base/cli (because sometimes it sits between main() and component-base/logs) should be. That makes sense to me.

I'll revive my alternative branch (a bit stale now because only rebased this one here) and swap the code. Then we can look at that alternative API and decide whether that's better.

Reply:
Cherry-picking https://github.com/pohly/kubernetes/commit/8d834839f87959a556bdfda8b9e23b5f81b816ab worked almost without conflicts. I've pushed a new branch with the initial commit removed and the last one replaced.

The feature gate implementation is now more complex. Previous code with access to a global instance:
```
$ git log -p -n1 log-configuration-feature-gates-2022-03-24  | diffstat
 api/v1/kube_features.go        |   59 +++++++++++++++++++++++++++++++
 api/v1/options.go              |   39 ++++++++++++++++++---
 api/v1/options_test.go         |   10 +++--
 api/v1/registry.go             |    6 +++
 api/v1/types.go                |   11 +++--
 api/v1/validate_test.go        |   43 ++++++++++++++++++++++-
 example/cmd/logger.go          |    2 +
 json/json.go                   |    5 ++
 json/register/validate_test.go |   76 +++++++++++++++++++++++++++++++++++++++++
 9 files changed, 237 insertions(+), 14 deletions(-)
```

Now:
```
$ git log -p -n1 log-configuration-feature-gates-2022-03-24-no-default | diffstat
 cmd/kube-apiserver/app/server.go                                      |    2 
 cmd/kube-controller-manager/app/controllermanager.go                  |    2 
 cmd/kube-scheduler/app/server.go                                      |    3 
 cmd/kubelet/app/server.go                                             |    2 
 pkg/features/kube_features.go                                         |    2 
 pkg/kubelet/apis/config/validation/validation.go                      |    2 
 staging/src/k8s.io/component-base/featuregate/feature_gate.go         |   14 ++
 staging/src/k8s.io/component-base/logs/api/v1/features.go             |   57 ++++++++
 staging/src/k8s.io/component-base/logs/api/v1/features_test.go        |   47 +++++++
 staging/src/k8s.io/component-base/logs/api/v1/options.go              |   51 ++++++-
 staging/src/k8s.io/component-base/logs/api/v1/options_test.go         |   12 +
 staging/src/k8s.io/component-base/logs/api/v1/registry.go             |    7 +
 staging/src/k8s.io/component-base/logs/api/v1/types.go                |   11 +
 staging/src/k8s.io/component-base/logs/api/v1/validate_test.go        |   35 +++++
 staging/src/k8s.io/component-base/logs/example/cmd/logger.go          |    9 +
 staging/src/k8s.io/component-base/logs/json/json.go                   |    5 
 staging/src/k8s.io/component-base/logs/json/register/register_test.go |   65 +++++++++-
 17 files changed, 296 insertions(+), 30 deletions(-)
```

Note that this is with a more forgiving implementation that doesn't panic when the passed in FeatureGate doesn't have the feature. That could be changed to panic because all callers should do the necessary registration, but it seems unnecessary to impose registration as a hard requirement on users of the package when there is a reasonable fallback.

The API change in this PR isn't too bad because not many places use LoggingConfiguration.ValidateAndApply.

Where it gets worse is in the implementation of the contextual logging feature gate, because there component-base/cli.Run calls component-base/logs.InitLogs and that is where the new "ContextualLogging" feature gate gets checked: https://github.com/pohly/kubernetes/commit/8d834839f87959a556bdfda8b9e23b5f81b816ab

@liggitt: please have a look at that commit and the last commit of this PR and let me know whether this is what you prefer or whether you see a simpler way of doing this.

Reply:
Speaking of contextual logging, getting that feature completed is more important than this PR here. Can we continue the discussion around feature gates to the PR that I just created for that? See https://github.com/kubernetes/kubernetes/pull/108995.